src/Init/Data/Json/Basic.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2019 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Gabriel Ebner, Marc Huisinga
+-/
+prelude
+import Init.System.IO
+import Init.Data.String 
+import Init.Data.Int
+
+-- mantissa * 10^-exponent
+structure JsonNumber :=
+(mantissa : Int) (exponent : Nat)
+
+protected def JsonNumber.decEq : ∀ (a b : JsonNumber), Decidable (a = b)
+| ⟨m1, e1⟩, ⟨m2, e2⟩ =>
+  match decEq m1 m2 with
+  | isTrue hm =>
+    match decEq e1 e2 with
+    | isTrue he  => isTrue (hm ▸ he ▸ rfl)
+    | isFalse he => isFalse (fun h => JsonNumber.noConfusion h (fun hm he2 => he he2))
+  | isFalse hm => isFalse (fun h => JsonNumber.noConfusion h (fun hm2 he => hm hm2))
+
+instance JsonNumber.DecidableEq : DecidableEq JsonNumber := 
+JsonNumber.decEq
+
+namespace JsonNumber
+
+protected def fromNat (n : Nat) : JsonNumber := ⟨n, 0⟩
+protected def fromInt (n : Int) : JsonNumber := ⟨n, 0⟩
+
+instance natToJsonNumber : HasCoe Nat JsonNumber := ⟨JsonNumber.fromNat⟩
+instance intToJsonNumber : HasCoe Int JsonNumber := ⟨JsonNumber.fromInt⟩
+
+private def countDigitsAux : Nat → Nat → Nat → Nat
+| 0  , _, _        => 0
+| i+1, n, digits => 
+  if n ≤ 9 then
+    digits
+  else
+    countDigitsAux i (n/10) (digits+1)
+
+private def countDigits (n : Nat) : Nat :=
+countDigitsAux n n 1
+
+protected def toString : JsonNumber → String
```

Comment:
Since I've been mentioned on this PR, let me add my mustard.  The reason I added a separate `JsonNumber` type in the first place (and not using rationals or floating point numbers) is for safe round-tripping of JSON numbers.  Floats have both a limit on the absolute value as well as the precision.  This leads to interesting bugs when you transfer large IDs via JSON: they are implicitly rounded (without any warning).

I think it is a good idea to make this rounding explicit, if the user actually wants it.  Since we already have arbitrary-size integers, this is easy to do.

Reply:
I'd also prefer to keep it this way for the reasons that Gabriel mentioned. 

In fact, I even considered to make the JsonNumber type slightly weirder, because currently ```mantissa``` is exponential in the input exponent. I decided not to bother for now, since other issues like the deep recursion one are much more glaring if one was to use the parser in user-facing code.

Reply:
Ah, I missed that this code is part of the roundtrip